examples/hnas/lstm_example.py

```diff
@@ -0,0 +1,504 @@
+        else:
+            try:
+                nonlinearity = {'RNN_TANH': 'tanh', 'RNN_RELU': 'relu'}[rnn_type]
+            except KeyError:
+                raise ValueError("""An invalid option for `--model` was supplied,
+                                 options are ['LSTM', 'GRU', 'RNN_TANH' or 'RNN_RELU']""")
+            self.rnn = nn.RNN(ninp, nhid, nlayers, nonlinearity=nonlinearity, dropout=dropout)
+        self.decoder = nn.Linear(nhid, ntoken)
+
+        # Optionally tie weights as in:
+        # "Using the Output Embedding to Improve Language Models" (Press & Wolf 2016)
+        # https://arxiv.org/abs/1608.05859
+        # and
+        # "Tying Word Vectors and Word Classifiers: A Loss Framework for Language Modeling" (Inan et al. 2016)
+        # https://arxiv.org/abs/1611.01462
+        if tie_weights:
+            if nhid != ninp:
+                raise ValueError('When using the tied flag, nhid must be equal to emsize')
+            self.decoder.weight = self.encoder.weight
+
+        self.init_weights()
+
+        self.rnn_type = rnn_type
+        self.nhid = nhid
+        self.nlayers = nlayers
+
+    def init_weights(self):
+        initrange = 0.1
+        self.encoder.weight.data.uniform_(-initrange, initrange)
+        self.decoder.bias.data.zero_()
+        self.decoder.weight.data.uniform_(-initrange, initrange)
+
+    def forward(self, input, hidden):
+        emb = self.drop(self.encoder(input))
+        output, hidden = self.rnn(emb, hidden)
+        output = self.drop(output)
+        decoded = self.decoder(output)
+        return decoded, hidden
+
+    def init_hidden(self, bsz):
+        weight = next(self.parameters())
+        if self.rnn_type == 'LSTM':
+            return (weight.new_zeros(self.nlayers, bsz, self.nhid),
+                    weight.new_zeros(self.nlayers, bsz, self.nhid))
+        else:
+            return weight.new_zeros(self.nlayers, bsz, self.nhid)
+
+
+def repackage_hidden(h):
+    """Wraps hidden states in new Tensors, to detach them from their history."""
+
+    if isinstance(h, torch.Tensor):
+        return h.detach()
+    else:
+        return tuple(repackage_hidden(v) for v in h)
+
+
+def make_run_wikitext2(cli_args):
+
+    @ag.args(lr=ag.space.Real(lower=1, upper=50, log=True),
+             dropout=ag.space.Real(lower=0, upper=0.99),
+             batch_size=ag.space.Int(lower=8, upper=128),
+             clip=ag.space.Real(lower=0.1, upper=2),
+             lr_factor=ag.space.Real(lower=1, upper=100, log=True),
+             epochs=cli_args.epochs,
+             num_gpus=cli_args.num_gpus)
+    def run_wikitext2(args, reporter, **kwargs):
+        print(args, kwargs)
+
+        model_type = "rnn"
+        data = 'wikitext-2/'
+
+        emsize = 1500
+        nhid = 1500
+        nlayers = 2
+        eval_batch_size = 10
+        bptt = 35
+
+        tied = True
+        seed = np.random.randint(10000)
```

Comment:
Do we want to fix seed if it is provided in args? Otherwise this one will harm reproducibility.